### PR TITLE
Add additional phase for timeouts

### PR DESCRIPTION
**Pull request by schrodit**
**What this PR does / why we need it**:
This PR adds an additional phase called "Timeout" which applied if a argo node exceeds their `activeDeadlineSeconds`.

**Special notes for your reviewer**:
Should we add the `finishedAt` and `duration` according to the activeDeadlineSeconds so that finishedAt = startTime + duration and duration = activeDeadlineSeconds.

This would mean that we have to add additional information to the status (activeDeadlineSeconds of the TestDefinitiobn) as not all information of the TestDefinition is available in the status.
**Release note**:

```improvement user
Add additional phase "Timeout" for test that exceeded the `activeDeadlineSeconds`
```

### CODE REVIEW

**Comment by schrodit** [comment]
Added the duration to the status of the testrun so that now the "real" duration is shown in the status and in elasticsearch

File: pkg/controller/testrun_control_update.go [hunk 1]
```diff
@@ -104,6 +108,17 @@ func updateStepsStatus(tr *tmv1beta1.Testrun, wf *argov1.Workflow) {
 				continue
 			}
 
+			if strings.Contains(argoNodeStatus.Message, argoctrl.ErrDeadlineExceeded.Error()) {
+				completionTime := metav1.NewTime(stepStatus.StartTime.Add(time.Duration(*stepStatus.TestDefinition.ActiveDeadlineSeconds) * time.Second))
```
Comment (OlegLoewen) [inline pkg/controller/testrun_control_update.go]: Maybe decouple a little bit for better reading?
```suggestion
				testDurationMs := time.Duration(*stepStatus.TestDefinition.ActiveDeadlineSeconds) * time.Second
				completionTime := metav1.NewTime(stepStatus.StartTime.Add(testDurationMs))
```

File: pkg/testmachinery/testdefinition/testdefinition.go [hunk 2]
```diff
@@ -52,13 +52,16 @@ func New(def *tmv1beta1.TestDefinition, loc Location, fileName string) *TestDefi
 	if def.Spec.Image == "" {
 		def.Spec.Image = testmachinery.BASE_IMAGE
 	}
+	if def.Spec.ActiveDeadlineSeconds == nil {
+		def.Spec.ActiveDeadlineSeconds = &activeDeadlineSeconds
```
Comment (OlegLoewen) [inline pkg/testmachinery/testdefinition/testdefinition.go]:
```suggestion
		def.Spec.ActiveDeadlineSeconds = &defaultActiveDeadlineSeconds
```

File: pkg/util/util.go [hunk 3]
```diff
@@ -45,7 +45,7 @@ func MaxTimeExceeded(startTime time.Time, maxWaitTime int64) bool {
 
 // Completed checks if the testrun is in a completed phase
 func Completed(phase argov1.NodePhase) bool {
-	if phase == argov1.NodeSucceeded || phase == argov1.NodeFailed || phase == argov1.NodeError || phase == argov1.NodeSkipped {
+	if phase == tmv1beta1.PhaseStatusSuccess || phase == tmv1beta1.PhaseStatusFailed || phase == tmv1beta1.PhaseStatusError || phase == tmv1beta1.PhaseStatusSkipped || phase == tmv1beta1.PhaseStatusTimeout {
```
Comment (OlegLoewen) [inline pkg/util/util.go]: Optional/Idea: as discussed maybe it's good to introduce a list.contains utility function to make this line more compact and readable? (In combination with a dot import of tmv1beta1)
```suggestion
    if stringInSlice(phase, []string{PhaseStatusSuccess, PhaseStatusFailed, PhaseStatusError, PhaseStatusSkipped, PhaseStatusTimeout}) {
func stringInSlice(a string, list []string) bool {
    for _, b := range list {
        if b == a {
            return true
        }
    }
    return false
}
```

File: pkg/testmachinery/testdefinition/testdefinition.go (hunk 2 above)
Comment (schrodit) [reply pkg/testmachinery/testdefinition/testdefinition.go]: done

File: pkg/controller/testrun_control_update.go (hunk 1 above)
Comment (schrodit) [reply pkg/controller/testrun_control_update.go]: done

File: pkg/util/util.go (hunk 3 above)
Comment (schrodit) [reply pkg/util/util.go]: Discussed offline

**Comment by schrodit** [comment]
@OlegLoewen @dguendisch added an integration for the new phase and fixed dependency errors.
can you please rereview?

**Comment by schrodit** [comment]
Problem was I needed another package for the pod message which resulted in a need to update some additional packages.
But I try to separate them next time:)

**Pull request merged**